File: leetcode/437-path-sum-iii/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	1st approach:
	- memorize the sum of all subpaths on each node
	- TLE
	Time O(n^2)
	Space...a lot
	14jan2019
*/
func pathSum_(root *TreeNode, sum int) int {
	res := 0
	var dfs func(node *TreeNode, agg []SumNode)
	dfs = func(node *TreeNode, agg []SumNode) {
		if node == nil {
			return
		}
		newAgg := []SumNode{}
		for i := 0; i < len(agg); i++ {
			newPath := []int{}
			newPath = append(newPath, agg[i].Paths...)
			newPath = append(newPath, node.Val)
			newSum := agg[i].Sum + node.Val
			newAgg = append(newAgg, SumNode{newPath, newSum})
			if newSum == sum {
				res++
			}
		}
		newAgg = append(newAgg, SumNode{[]int{node.Val}, node.Val})
		if node.Val == sum {
			res++
		}
		dfs(node.Left, newAgg)
		dfs(node.Right, newAgg)
	}
	dfs(root, []SumNode{})
	return res
}

type SumNode struct {
	Paths []int
	Sum   int
}

/*
	2nd approach:
	- optimize the above one
	e.g. for path [10, 5, 2, 1], for each node 10 -> [10], 5 -> [15,5], 2 -> [17,7,2], 1 -> [18,8,3,1]
	Time		O(n^2)
	Space		O(n^2)
	24ms beats 28.57%
	14jan2019
*/
func pathSum(root *TreeNode, sum int) int {
	res := 0
	var dfs func(node *TreeNode, agg []int)
	dfs = func(node *TreeNode, agg []int) {
		if node == nil {
			return
		}
		newAgg := []int{}
		for i := 0; i < len(agg); i++ {
			temp := agg[i] + node.Val
			newAgg = append(newAgg, temp)
			if temp == sum {
				res++
			}
		}
		newAgg = append(newAgg, node.Val)
		if node.Val == sum {
			res++
		}
		dfs(node.Left, newAgg)
		dfs(node.Right, newAgg)
	}
	dfs(root, []int{})
	return res
}

func main() {
	// 			10
	//		5		-3
	// 3	 2	  11
	//3 -2  1
	root := &TreeNode{10,
		&TreeNode{5,
			&TreeNode{3,
				&TreeNode{3,
					nil,
					nil,
				},
				&TreeNode{-2,
					nil,
					nil,
				},
			},
			&TreeNode{2,
				nil,
				&TreeNode{1,
					nil,
					nil,
				},
			},
		},
		&TreeNode{-3,
			nil,
			&TreeNode{11,
				nil,
				nil,
			},
		},
	}
	fmt.Println(pathSum(root, 8))
}

File: leetcode/437-path-sum-iii/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


"""
    1st approach: recursive dfs with prefix sum
    
    Time    O(n*h) h: average height of the tree
    Space   O(h)
    408 ms, faster than 55.24%
    1apr2019
"""


class Solution(object):

    def __init__(self):
        self.count = 0

    def pathSum(self, root, target):
        self.helper(root, [], target)
        return self.count

    def helper(self, node, prefixArr, target):
        if node == None:
            return
        newArr = []
        # add up the current value to the items in the prefix sum array
        for s in prefixArr:
            temp = s + node.val
            newArr.append(temp)
            if temp == target:
                self.count += 1
        # add the current value into the prefix sum array
        newArr.append(node.val)
        if node.val == target:
            self.count += 1
        # traverse the children
        self.helper(node.left, newArr, target)
        self.helper(node.right, newArr, target)


"""
    2nd approach: iterative dfs with prefix sum
    
    Time    O(n*h) h: average height of the tree
    Space   O(h)
    336 ms, faster than 58.38%
    1apr2019
"""


class StackItem(object):
    def __init__(self, node, prefixArr=[]):
        self.node = node
        self.prefixArr = prefixArr


class Solution(object):

    def pathSum(self, root, target):
        if root == None:
            return 0
        count = 0
        stack = []
        stack.append(StackItem(root))
        while len(stack) > 0:
            pop = stack.pop()

            node = pop.node
            prefixArr = pop.prefixArr

            # add up the current value to the items in the prefix sum array
            newArr = []
            for s in prefixArr:
                temp = s + node.val
                newArr.append(temp)
                if temp == target:
                    count += 1
            # add the current value into the prefix sum array
            newArr.append(node.val)
            if node.val == target:
                count += 1

            # traverse down to its children
            if node.left != None:
                stack.append(StackItem(node.left, newArr))
            if node.right != None:
                stack.append(StackItem(node.right, newArr))
        return count
